### **Author : Midhir Nambiar**

### **Task : Heart Disease Prediction using Regression**

Problem Statement - Given a dataset in which we need to build a predictive model which would predict whether the person has a defective heart(heart disease) or not.

Outline:

1. Importing Dataset
2. Data Cleaning & Data Preparation
3. Train-Test Split
4. Model Building & Training
5. Model Evaluation
6. Build a Predictive Model


In [17]:
## Importing Libraries for Analysis
import numpy as np
import pandas as pd

# Import Libraries for Train-Test Split
from sklearn.model_selection import train_test_split

# Import Libraries for Model Building
from sklearn.linear_model import LogisticRegression

# Import Libraries for Accuracy Score
from sklearn.metrics import accuracy_score

# Import Libraries to Supress Warnings
import warnings
warnings.filterwarnings('ignore')

### **Importing Dataset**

In [2]:
heart_df = pd.read_csv('/content/heart.csv')

In [3]:
heart_df.head()

age  sex  cp  trestbps  chol  fbs  ...  exang  oldpeak  slope  ca  thal  target
0   63    1   3       145   233    1  ...      0      2.3      0   0     1       1
1   37    1   2       130   250    0  ...      0      3.5      0   0     2       1
2   41    0   1       130   204    0  ...      0      1.4      2   0     2       1
3   56    1   1       120   236    0  ...      0      0.8      2   0     2       1
4   57    0   0       120   354    0  ...      1      0.6      2   0     2       1

[5 rows x 14 columns]

In [4]:
heart_df.tail()

age  sex  cp  trestbps  chol  fbs  ...  exang  oldpeak  slope  ca  thal  target
298   57    0   0       140   241    0  ...      1      0.2      1   0     3       0
299   45    1   3       110   264    0  ...      0      1.2      1   0     3       0
300   68    1   0       144   193    1  ...      0      3.4      1   2     3       0
301   57    1   0       130   131    0  ...      1      1.2      1   1     3       0
302   57    0   1       130   236    0  ...      0      0.0      1   1     2       0

[5 rows x 14 columns]

### **Data Cleaning & Preparation**

In [5]:
print(heart_df.shape)

(303, 14)


In [6]:
heart_df.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [7]:
heart_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [8]:
heart_df.describe()

age         sex          cp  ...          ca        thal      target
count  303.000000  303.000000  303.000000  ...  303.000000  303.000000  303.000000
mean    54.366337    0.683168    0.966997  ...    0.729373    2.313531    0.544554
std      9.082101    0.466011    1.032052  ...    1.022606    0.612277    0.498835
min     29.000000    0.000000    0.000000  ...    0.000000    0.000000    0.000000
25%     47.500000    0.000000    0.000000  ...    0.000000    2.000000    0.000000
50%     55.000000    1.000000    1.000000  ...    0.000000    2.000000    1.000000
75%     61.000000    1.000000    2.000000  ...    1.000000    3.000000    1.000000
max     77.000000    1.000000    3.000000  ...    4.000000    3.000000    1.000000

[8 rows x 14 columns]

In [9]:
heart_df['target'].value_counts()

1    165
0    138
Name: target, dtype: int64


Conclusion : **165** people have a heart disease.





**Splitting the features and target**

In [10]:
X = heart_df.drop(columns = 'target' , axis = 1)
Y = heart_df['target']

In [11]:
print(X)

     age  sex  cp  trestbps  chol  ...  exang  oldpeak  slope  ca  thal
0     63    1   3       145   233  ...      0      2.3      0   0     1
1     37    1   2       130   250  ...      0      3.5      0   0     2
2     41    0   1       130   204  ...      0      1.4      2   0     2
3     56    1   1       120   236  ...      0      0.8      2   0     2
4     57    0   0       120   354  ...      1      0.6      2   0     2
..   ...  ...  ..       ...   ...  ...    ...      ...    ...  ..   ...
298   57    0   0       140   241  ...      1      0.2      1   0     3
299   45    1   3       110   264  ...      0      1.2      1   0     3
300   68    1   0       144   193  ...      0      3.4      1   2     3
301   57    1   0       130   131  ...      1      1.2      1   1     3
302   57    0   1       130   236  ...      0      0.0      1   1     2

[303 rows x 13 columns]


In [12]:
print(Y)

0      1
1      1
2      1
3      1
4      1
      ..
298    0
299    0
300    0
301    0
302    0
Name: target, Length: 303, dtype: int64


### **Train-Test Split**

**NOTE** : stratify distributes 1 and 0 equally.

In [13]:
X_train , X_test , Y_train , Y_test = train_test_split(X , Y , test_size = 0.3 , stratify = Y , random_state = 2)

In [14]:
print(X.shape , X_train.shape , X_test.shape)

(303, 13) (212, 13) (91, 13)


### **Model Building & Training**

In [33]:
model = LogisticRegression()

Training our model with training data

In [16]:
model.fit(X_train , Y_train)

LogisticRegression()

### **Model Evaluation**

In [20]:
# accuracy on training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction , Y_train)

In [24]:
print('Accuracy on training data : ', training_data_accuracy*100)

Accuracy on training data :  84.43396226415094


In [25]:
# accuracy on test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction , Y_test)

In [26]:
print('Accuracy on test data : ', test_data_accuracy*100)

Accuracy on test data :  85.71428571428571


### **Build a Predictive Model**

In [31]:
input_data = (41,0,1,130,204,0,0,172,0,1.4,2,0,2)

# change the input_data to a numpy array
input_data_as_numpy_array = np.asarray(input_data)

# reshape the numpy array as we are predicting for only on instance
input_data_reshaped = input_data_as_numpy_array.reshape(1, -1)

prediction = model.predict(input_data_reshaped)
print(prediction)

if (prediction[0] == 0):
  print('The person does not have an heart disease')
else:
  print('The person has a heart disease')

[1]
The person has a heart disease


**Conclusion**: Hence we were successfull in building a model which would predict whether a person has a heart disease or not.

### **THANK YOU**